source/core/module_coreHelpers.js

```diff
@@ -75,33 +104,68 @@
                     str += String.fromCharCode(u0);
                     continue;
                 }
+
+                // Look ahead to validate the UTF-8 structure
+                if ((u0 & 0xE0) === 0xC0) {
+                    if (idx >= endIndex || (u8Array[idx] & 0xC0) !== 0x80) {
+                        str += '\uFFFD';
+                        continue;
+                    }
+                } else if ((u0 & 0xF0) === 0xE0) {
```

Comment:
Maybe adding helper functions to end up with something like:
`if(bytesRequired(u0) == 1) { ... }`